### PR TITLE
add Encoder and Decoder

### PR DESCRIPTION
**Pull request by kevin-hanselman**
A port of ghodss/yaml#39

### CODE REVIEW

**Comment by dims** [comment]
Hey @kevin-hanselman thanks for doing this. Could i persuade you to consider adding some benchmarks? (the linked PR hints at "memory footprint" gains ... so `-benchmem`?)

**Comment by dims** [comment]
/assign @liggitt @deads2k 

folks, can you please peek?

**Comment by kevin-hanselman** [comment]
Added a (very) basic benchmark test. Here's the results on my machine:

This branch:
```
goos: linux
goarch: amd64
pkg: sigs.k8s.io/yaml
cpu: Intel(R) Core(TM) i3-7100 CPU @ 3.90GHz
BenchmarkMarshal-4        153171              7745 ns/op            9530 B/op         50 allocs/op
BenchmarkUnmarshal-4       36927             32184 ns/op           11413 B/op        151 allocs/op
PASS
ok      sigs.k8s.io/yaml        2.788s
```

master:
```
goos: linux
goarch: amd64
pkg: sigs.k8s.io/yaml
cpu: Intel(R) Core(TM) i3-7100 CPU @ 3.90GHz
BenchmarkMarshal-4        249537              4672 ns/op            8937 B/op         42 allocs/op
BenchmarkUnmarshal-4       65419             18461 ns/op           10891 B/op        141 allocs/op
PASS
ok      sigs.k8s.io/yaml        2.615s
```

I'll need to think more about these results and code before drawing any conclusions.

**Comment by kevin-hanselman** [comment]
@dims Anything else I can do to move this work along?

**Comment by dims** [comment]
@deads2k @sttts @deads2k thoughts please!

cc @kubernetes-sigs/kubernetes-sig-api-machinery-reviewers

File: yaml.go [hunk 1]
```diff
@@ -11,20 +11,98 @@ import (
...
+// Encode serializes a value into YAML and writes it using the Encoder's
+// io.Writer.
+func (e *Encoder) Encode(o interface{}) error {
+	pipeReader, pipeWriter := io.Pipe()
+	defer pipeReader.Close()
+
+	var intermediate interface{}
+	jsonEnc := json.NewEncoder(pipeWriter)
+	// We are using yaml.Decoder here (instead of json.Decoder) because the Go
+	// JSON library doesn't try to pick the right number type (int, float,
```
Comment (liggitt) [inline yaml.go:33]: json.NewDecoder(...).UseNumber() will preserve numeric input literally (https://pkg.go.dev/encoding/json#Decoder.UseNumber) ... would that enable translating with fidelity? how does the json decoder compare to the yaml decoder in efficiency?

File: yaml.go [hunk 2]
```diff
@@ -11,20 +11,98 @@ import (
...
+	// etc.) when unmarshalling to interface{}; it always picks float64.
+	// go-yaml preserves the number type.
+	yamlDec := yaml.NewDecoder(pipeReader)
+	yamlEnc := yaml.NewEncoder(e.w)
+
+	var jsonErr error
+	go func() {
```
Comment (liggitt) [inline yaml.go:40]: spawning goroutines here and in Decode is a little hard to reason about... a few things I noticed/wondered

* if this panics, the calling program will exit instead of propagating the panic to the Encode() caller... I'd expect the panic to propagate instead
* is it better to start the json encoder async and the yaml decoder inline or vice-versa?
* is it better to leave the pipereader unbuffered or wrap in a buffer?

File: yaml.go [hunk 3]
```diff
@@ -11,20 +11,98 @@ import (
...
+		jsonErr = jsonEnc.Encode(o)
+		pipeWriter.Close()
+	}()
+	yamlErr := yamlDec.Decode(&intermediate)
+	if jsonErr != nil {
```
Comment (liggitt) [inline yaml.go:45]: I'm surprised this doesn't complain about races the same way

File: yaml.go (hunk 1 above)
Comment (liggitt) [reply yaml.go:33]: also, a unit test demonstrating some of the edge cases like this would be helpful in ensuring this isn't producing user-visible changes

other examples of weird edges in yaml encoding/decoding we don't want to accidentally modify behavior on (even if the current behavior is not ultimately desired/correct):
* unquoted yaml keys of `y`: https://github.com/kubernetes/kubernetes/issues/102308
* yaml parsing of unquoted values containing all digits starting with 0 which are not octal values: https://github.com/kubernetes/kubernetes/issues/82296

File: yaml.go (hunk 2 above)
Comment (kevin-hanselman) [reply yaml.go:40]: FWIW, this is adapted from the [official `io.Pipe` docs](https://pkg.go.dev/io#example-Pipe).

File: yaml.go [hunk 4]
```diff
@@ -11,20 +11,98 @@ import (
...
+	defer pipeReader.Close()
+
+	var intermediate interface{}
+	jsonEnc := json.NewEncoder(pipeWriter)
+	// We are using yaml.Decoder here (instead of json.Decoder) because the Go
+	// JSON library doesn't try to pick the right number type (int, float,
+	// etc.) when unmarshalling to interface{}; it always picks float64.
+	// go-yaml preserves the number type.
+	yamlDec := yaml.NewDecoder(pipeReader)
+	yamlEnc := yaml.NewEncoder(e.w)
```
Comment (lavalamp) [inline yaml.go:37]: This is doing an encode/decode/encode? Why? There has to be a better way than this? There should be some comment explaining the purpose?

**Comment by VictorLowther** [comment]
On Tue, Sep 7, 2021, 16:11 Daniel Smith ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In yaml.go
> <https://github.com/kubernetes-sigs/yaml/pull/59#discussion_r703833418>:
>
> > +}
> +
> +// Encode serializes a value into YAML and writes it using the Encoder's
> +// io.Writer.
> +func (e *Encoder) Encode(o interface{}) error {
> +	pipeReader, pipeWriter := io.Pipe()
> +	defer pipeReader.Close()
> +
> +	var intermediate interface{}
> +	jsonEnc := json.NewEncoder(pipeWriter)
> +	// We are using yaml.Decoder here (instead of json.Decoder) because the Go
> +	// JSON library doesn't try to pick the right number type (int, float,
> +	// etc.) when unmarshalling to interface{}; it always picks float64.
> +	// go-yaml preserves the number type.
> +	yamlDec := yaml.NewDecoder(pipeReader)
> +	yamlEnc := yaml.NewEncoder(e.w)
>
> This is doing an encode/decode/encode? Why? There has to be a better way
> than this? There should be some comment explaining the purpose?
>

Better is a fuzzy metric. I do not use this package in any performance
critical paths, it got the job done, and I have never had to touch it again.

**Comment by luxas** [comment]
Cross-referencing https://github.com/kubernetes-sigs/yaml/pull/61#issuecomment-920711786 with information about potential blockers / info to think about with regards to this PR.
